### ISSUE TITLE
[Train] Automatically choose number of workers

### ISSUE THREAD

**Issue by bveeramani** [ray-project/ray#21987]
### Search before asking

- [X] I had searched in the [issues](https://github.com/ray-project/ray/issues) and found no similar feature requirement.

### Description

Users must specify the number of workers in the `Trainer` constructor.

```
trainer = Trainer(num_workers=8, backend="torch")
```

We could add a feature that automatically chooses the optimal number of workers. It could look something like this.

```
trainer = Trainer(num_workers="auto", backend="torch")
```

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

**Comment by matthewdeng** [comment]
@Yard1 maybe related to Tune dynamic resources

**Comment by cszhu** [comment]
Hello! This P3 issue has seen no activity in at least a year. It will be closed in 2 weeks as part of ongoing cleanup efforts.

Please remove the `pending-cleanup` label if you believe this issue should remain open.

Thanks for contributing to Ray!
